### PR TITLE
Implement Sinsemilla primitives

### PR DESCRIPTION
**Pull request by str4d**
Test vectors will be added in a subsequent PR.

### CODE REVIEW

File: src/primitives/sinsemilla.rs
Comment (therealyingtong) [inline src/primitives/sinsemilla.rs:48]: should this be `pallas::Affine` instead?

File: src/primitives/sinsemilla.rs
Comment (dconnolly) [inline src/primitives/sinsemilla.rs:26]: Same here, should probably be `pallas::Affine`; `Extract_P` is defined as having a `P` element input which is an affine point:
image
image

Style opinion, as this is an open PR vs needing to open a new thing in `halo2`: naming all point types as explicitly *Point, such as `AffinePoint`, `ProjectivePoint`, etc, which can improve readability and human distinguishability of types. The [`jubjub` crate](https://docs.rs/jubjub/0.6.0/jubjub/) basically does this. 

I know in `halo2` a lot of these are generated by macros but the [names are chosen on macro invocation](https://github.com/zcash/halo2/blob/63083e418cdd836e0056dd55f6498ccf185af83f/src/pasta/curves.rs#L908) and renamed and exported in the [`pallas` module](https://github.com/zcash/halo2/blob/main/src/pasta/pallas.rs#L15).  This is of course not limited to or blocking on this PR but is a kind of naming / typing style thing we zcash engineers can discuss, if desired.

File: src/primitives/sinsemilla.rs
Comment (dconnolly) [reply src/primitives/sinsemilla.rs:48]: Agreed, the spec currently defines `P` as the group of points `(x,y)` on the short Weierstrass form of the curve.

image
image

File: src/primitives/sinsemilla.rs
Comment (dconnolly) [inline src/primitives/sinsemilla.rs:75]: `SinsemillaCommit` as an instance of `SinsemillaHashToPoint` also returns an affine point as an element of `P`

image
image
image

Comment (therealyingtong) [reply src/primitives/sinsemilla.rs:75]: yup, @str4d i've made a few changes to `primitives/sinsemilla.rs` to get it working with my WIP chip here: https://github.com/zcash/orchard/pull/25/files

File: src/primitives/sinsemilla.rs
Comment (str4d) [reply src/primitives/sinsemilla.rs:48]: `pallas::Point` and `pallas::Affine` are equivalent group elements by construction and have 1:1 conversion (enforced by the `group::Curve` and `group::prime::{PrimeCurve, PrimeCurveAffine}` traits). However, `pallas::Point: group::Curve` is the format that is efficient to perform operations on, and is intended to be what is passed around in-memory. `pallas::Affine` is only intended for encoding/decoding and accessing affine coordinates. Returning `pallas::Affine` from this API would force an unnecessary `projective -> affine -> projective` round-trip on every call to `sinsemilla::commit` or `sinsemilla::short_commit`.

File: src/primitives/sinsemilla.rs
Comment (str4d) [reply src/primitives/sinsemilla.rs:26]: Having this take `pallas::Affine` pushes a `point.to_affine()` call onto every caller, instead of locating them all inside this function. There are only a couple of callers though, so I don't overly mind.

Re: naming of `pallas::Point`, I opted in the `halo2` crate to leave the point representation out of the re-exported name for overall readability (and also because as above, it's irrelevant to the consumer). What mattered more to me was that I could write `pallas::Point` _everywhere_, so that it was clear that this is a Pallas point (rather than Vesta or some other curve); `pallas::ProjectivePoint` is almost twice as long.

`jubjub` is a slightly different case, because we need to represent both the curve type _and_ whether the point is constrained to the prime-order subgroup, so we already need to include more information in the name (in hindsight `jubjub::FullPoint` would have been a better name for this purpose than `jubjub::ExtendedPoint`). Pallas and Vesta are prime-order curves and don't have this problem.

File: src/primitives/sinsemilla.rs
Comment (dconnolly) [reply src/primitives/sinsemilla.rs:48]: I basically retract my comment, as I did basically the same thing in https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/sapling/commitment/pedersen_hashes.rs, even though the [Jubjub group element definitions in the spec](https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash) are also 'just' `(u, v)` etc. 🙈

I would love if the type was named `pallas::ProjectivePoint` or something. 😅

File: src/primitives/sinsemilla.rs
Comment (dconnolly) [reply src/primitives/sinsemilla.rs:26]: `pallas::ProjectivePoint` may be longer but it's not super long, and since there are multiple point types, vs a 'Point' and an 'Affine', it would be nice in terms of having the code document itself. Luckily Rust compiles away token names so we need not be precious on name lengths, as long we we don't turn into Java (😅 )

Comment (str4d) [reply src/primitives/sinsemilla.rs:26]: I'm only weakly against a rename (I like Dodger's "weakly" vs "strongly" for indicating preference), so I'll defer to others on this.

File: src/primitives/sinsemilla.rs
Comment (therealyingtong) [reply src/primitives/sinsemilla.rs:75]: i've reverted these changes; now calling `to_affine()` from the caller instead. :+1:

File: Cargo.toml
Comment (therealyingtong) [inline Cargo.toml:23]: was this change referencing the specific commit intentional?

File: src/primitives/sinsemilla.rs
Comment (daira) [reply src/primitives/sinsemilla.rs:48]: What @str4d said. Basically you're always allowed to use isomorphic representations.

File: src/primitives/sinsemilla.rs
Comment (daira) [reply src/primitives/sinsemilla.rs:26]: I like the current short names.

File: src/primitives/sinsemilla.rs
Comment (str4d) [inline src/primitives/sinsemilla.rs:17]:
```suggestion
fn lebs2ip_10(bits: &[bool]) -> u32 {
```

File: src/primitives/sinsemilla.rs
Comment (str4d) [inline src/primitives/sinsemilla.rs:54]:
```suggestion
    let S = |chunk: &[bool]| hasher_S(&lebs2ip_10(chunk).to_le_bytes());
```

**Review by ebfull — APPROVED** [review]

**Pull request merged**
